README.md
# jQuery Populaty - A Form Populator - [wbotelhos.com/populaty](http://wbotelhos.com/populaty)

jQuery Populaty is a plugin to populate form from a JSON.

## Version

```
@version        0.1.2
@since          2013-05-08
@author         Washington Botelho
@documentation  wbotelhos.com/populaty
@twitter        twitter.com/wbotelhos
```

## Required Files

+ jquery.populaty.js

## Options

```js
checkable  : ['checkbox', 'radio']             // Checkable fields.
exclude    : ':submit, :reset, :button'        // Field to be excluded from includeds.
include    : ':input'                          // Fields to be included on population.
json       : undefined                         // JSON data to populate the fields.
reset      : true                              // Reset or not the form before populate.
selectable : ['select-one', 'select-multiple'] // Selectable fields.
typeful    : [                                 // Typeful fields.
               'color',
               'date',
               'datetime',
               'datetime-local',
               'email',
               'hidden',
               'month',
               'number',
               'password',
               'range',
               'search',
               'tel',
               'text',
               'textarea',
               'time',
               'url',
               'week'
             ]
```

## Usage

```html
<form>
  <input name="number" />
</form>
```

```js
$('form').populaty({ json: { number: 1 } });
```

## Licence

The MIT License

## Donate

You can do it via [PayPal](https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=X8HEP2878NDEG&item_name=jQuery%20Populaty). Thanks! (:


changelog.md
# jQuery Populaty - A Form Populator - http://wbotelhos.com/populaty

## 0.1.2

### Fixes

+ Added some fields types of HTML 4 and 5 to typeful option.

## 0.1.1

### Fixes

+ When the value from json is boolean the value doesn't match with field's value.

## 0.1.0

### News

+ The JSON can be a String value or a Object value;
+ The population is based on fields inside de form, avoiding waste seek on all over JSON.
+ You can choose the field that will be candidate to be populated;
+ You can exclude field from the included one;
+ You can populate a form from a JSON;
+ You can reset the form before populate to have just the loaded data on form;
+ You can set a default JSON to be merged if your loaded one;
+ You can specify wich fields is checkable;
+ You can specify wich fields is selectables;
+ You can specify wich fields is typeful;


spec/lib/fixtury.js
clean = true;

function context(description, spec) {
  describe(description, spec);
};

var Fixtury = {
  init: function() {
  }, _checks: function(tag, data) {
    var type   = $(tag)[0].type,
        single = tag.indexOf(' />') !== -1;

    if (data.checked) {
      if (type === 'checkbox') {
        tag = tag.replace(' />', ' checked="checked" />');
      } else {
        $.error('You can check just checkbox element!');
      }
    } else if (data.selected) {
      if (type === 'radio') {
        tag = tag.replace(' />', ' selected="selected" />');
      } else if (tag.indexOf('<option') !== -1) {
        tag = tag.replace('>', ' selected="selected">');
      } else {
        $.error('You can select just radio and option element!');
      }
    }

    return tag;
  }, _data: function(options) {
    options = options || {};

    var times      = Fixtury._getTimes(options),
        html       = Fixtury._getHtml(options),
        checked    = Fixtury._getChecked(options),
        selected   = Fixtury._getSelected(options),
        opt        = Fixtury._normalize(options),
        attributes = Fixtury._parameterize(opt);

    return { attributes: attributes, html: html, checked: checked, selected: selected, times: times };
  }, _getChecked: function(options) {
    return options.checked || false;
  }, _getHtml: function(options) {
    var html    = options.html || '',
        content = '';

    if (typeof html === 'object') {
      for (var i = 0; i < html.length; i++) {
        content += html[i];
      }
    } else {
      content = html;
    }

    return content;
  }, _getSelected: function(options) {
    return options.selected || false;
  }, _getTimes: function(options) {
    return options.times || 1;
  }, _normalize: function(options) {
    delete options.times;
    delete options.html;
    delete options.selected;
    delete options.checked;

    return options;
  }, _parameterize: function(options) {
    var content = '';

    for (var option in options) {
      content += option + '="' + options[option] + '" ';
    }

    return content.replace(/\s$/, '');
  }, _repeat: function(tag, data) {
    var html = '';

    for (var i = 0; i < data.times; i++) {
      html += Fixtury._checks(tag, data).replace(/{index}/g, i + 1);
    }

    return html;
  }, _verify: function(options) {
    if (options && options.type) {
      $.error('You cannot set the "type" using an alias!');
    }
  }, append: function(html) {
    return $(html).appendTo('.fixtury');
  }, checkbox: function(options) {
    Fixtury._verify(options);

    return Fixtury.input(options, 'checkbox');
  }, clear: function() {
    if (clean) {
      $('.fixtury').empty();
    }
  }, double: function(options, name) {
    var data = Fixtury._data(options),
        tag  = '<' + name + ' ' + data.attributes + '>' + data.html + '</' + name + '>';

    return Fixtury._repeat(tag.replace(' >', '>'), data);
  }, fieldset: function(options) {
    return Fixtury.double(options, 'fieldset');
  }, form: function(options) {
    return Fixtury.double(options, 'form');
  }, hidden: function(options) {
    Fixtury._verify(options);

    return Fixtury.input(options, 'hidden');
  }, html: function(html) {
    $('.fixtury').html(html);
  }, input: function(options, type) {
    options = options || {};

    if (type) {
      options['type'] = type;
    }

    return Fixtury.single(options, 'input');
  }, label: function(options) {
    return Fixtury.double(options, 'label');
  }, legend: function(options) {
    return Fixtury.double(options, 'legend');
  }, option: function(options) {
    return Fixtury.double(options, 'option');
  }, password: function(options) {
    Fixtury._verify(options);

    return Fixtury.input(options, 'password');
  }, radio: function(options) {
    Fixtury._verify(options);

    return Fixtury.input(options, 'radio');
  }, select: function(options) {
    return Fixtury.double(options, 'select');
  }, single: function(options, name) {
    var data = Fixtury._data(options),
        tag  = '<' + name + ' ' + data.attributes + ' />';

    return Fixtury._repeat(tag.replace('  />', ' />'), data);
  }, submit: function(options) {
    Fixtury._verify(options);

    return Fixtury.input(options, 'submit');
  }, text: function(options) {
    Fixtury._verify(options);

    return Fixtury.input(options, 'text');
  }, textarea: function(options) {
    return Fixtury.double(options, 'textarea');
  }
};


spec/lib/spec.helper.js
// $.fn.populaty.defaults.debug = true;

// TODO: test the 'file' type.
var typefulTypes = ['color'];

function pathFromName(name) {
  var form = Fixtury.append(Fixtury.form({ html: Fixtury.text({ 'name': name }) }));

  form.populaty();

  return form.populaty('_extract', form.children('input')[0]);
};

function text(name, json) {
  var form = Fixtury.append(Fixtury.form({ html: Fixtury.text({ name: name }) }));

  form.populaty({ json: json });

  return form.children('input');
};

function textarea(name, json) {
  var form = Fixtury.append(Fixtury.form({ html: Fixtury.textarea({ name: name }) }));

  form.populaty({ json: json });

  return form.children('textarea');
};

function select(name, options, json, isMultiple) {
  isMultiple = isMultiple || false;

  var form = Fixtury.append(Fixtury.form({ html: Fixtury.select({ name: name, multiple: isMultiple, html: options }) }));

  form.populaty({ json: json });

  return form.children('select');
};

function hidden(name, json) {
  var form = Fixtury.append(Fixtury.form({ html: Fixtury.hidden({ name: name }) }));

  form.populaty({ json: json });

  return form.children('input');
};

function multiple(name, options, json) {
  select(name, options, json, true);
};

function checkbox(options, json) {
  var form = Fixtury.append(Fixtury.form({ html: options }));

  form.populaty({ json: json });

  return form;
};

function radio(options, json) {
  var form = Fixtury.append(Fixtury.form({ html: options }));

  form.populaty({ json: json });

  return form;
};


lib/jquery.populaty.js
/*!
 * jQuery Populaty - A Form Populator
 * ----------------------------------------------------------------------
 *
 * jQuery Populaty is a plugin to populate form from a JSON.
 *
 * Licensed under The MIT License
 *
 * @version        0.1.2
 * @since          2013-05-08
 * @author         Washington Botelho
 * @documentation  wbotelhos.com/populaty
 *
 * ----------------------------------------------------------------------
 *
 *  <form>
 *    <input name="number" />
 *  </form>
 *
 *  $('form').populaty({ json: { number: 1 } });
 *
 */

;(function($) {

  var methods = {
    init: function(settings) {
      return this.each(function() {
        if (settings && settings.json) {
          $.extend(settings.json, $.fn.populaty.defaults.json);
        }

        this.opt = $.extend({}, $.fn.populaty.defaults, settings);

        var that = this;

        if (!that.opt.json) {
          return;
        }

        if (that.opt.reset) {
          that.reset();
        }

        that.fields   = methods._fields.call(that);
        that.opt.json = methods._json.call(that);

        that.fields.each(function() {
          methods._seek.call(that, this);
        });

        $(that).data({ 'settings': that.opt, 'populaty': true });
      });
    }, _apply: function(field, value) {
      if (methods._is(field, this.opt.typeful)) {
        field.value = value;
      } else if (methods._is(field, this.opt.checkable)) {
        methods._markWith('checked', field, value);
      } else if (methods._is(field, this.opt.selectable)) {
        var options = $(field).children();

        for (var i = 0; i < options.length; i++) {
          methods._markWith('selected', options[i], value);
        }
      }

      methods._triggerEvents(field);
    }, _compact: function(array) {
      for (var i = 0; i < array.length; i++) {
        if (!array[i]) {
          array.splice(i--, 1);
        }
      }

      return array;
    }, _extract: function(field) {
      var paths  = (field.name || '').split('.'),
          result = []

      for (var item in paths) {
        var attributes = paths[item].split(/\[([\w]*)\]/);

        attributes = methods._compact(attributes);

        result = result.concat(attributes);
      }

      return result.length ? result : paths;
    }, _fields: function() {
      return $(this).find(this.opt.include).not(this.opt.exclude);
    }, _is: function(field, types) {
      return $.inArray(field.type, types) >= 0;
    }, _json: function() {
      var json = this.opt.json;

      return (typeof json === 'string') ? JSON.parse(json) : json;
    }, _markWith: function(mark, field, value) {
      value = value && value.toString();

      if (field.value == value) {
        var el = $(field);

        if (field.type !== 'select-multiple') {
          el.val(value);
        }

        el.prop(mark, mark);
      }
    }, _seek: function(field) {
      var that = this,
          path = methods._extract(field);

      (function seek(value, path) {
        for (var i = 0; i < path.length; i++) {
          value = value[path[i]];

          if (!value) {
            break;
          }

          if (typeof value === 'object') {
            if (value instanceof Array) {
              for (var item in value) {
                var subValue = value[item],
                    skip     = /[\d]/.test(path[i + 1]) ? 2 : 1,
                    subPath  = path.slice(i + skip);

                if (typeof subValue === 'object') {
                  seek(subValue, subPath)
                } else {
                  methods._apply.call(that, field, subValue);
                }
              }

              break;
            }
          } else {
            methods._apply.call(that, field, value);
          }
        }
      })(that.opt.json, path);
    }, _triggerEvents: function(field) {
      var el     = $(field),
          events = el.data('events');

      if (events) {
        if (events.click) {
          el.triggerHandler('click');
        }

        if (events.change) {
          el.triggerHandler('change');
        }
      }

      var onclick  = field.onclick,
          onchange = field.onchange;

      if (onclick) {
        onclick.call(field);
      }

      if (onchange) {
        onchange.call(field);
      }
    }
  };

  $.fn.populaty = function(method) {
    if (methods[method]) {
      return methods[method].apply(this, Array.prototype.slice.call(arguments, 1));
    } else if (typeof method === 'object' || !method) {
      return methods.init.apply(this, arguments);
    } else {
      $.error('Method ' + method + ' does not exist!');
    }
  };

  $.fn.populaty.defaults = {
    checkable  : ['checkbox', 'radio'],
    exclude    : ':submit, :reset, :button',
    include    : ':input',
    json       : undefined,
    reset      : true,
    selectable : ['select-one', 'select-multiple'],
    typeful    : [
                   'color',
                   'date',
                   'datetime',
                   'datetime-local',
                   'email',
                   'hidden',
                   'month',
                   'number',
                   'password',
                   'range',
                   'search',
                   'tel',
                   'text',
                   'textarea',
                   'time',
                   'url',
                   'week'
                 ]
  };

})(jQuery);
